module/os/linux/spl/spl-taskq.c

```diff
@@ -1255,12 +1284,101 @@ module_param_call(spl_taskq_kick, param_set_taskq_kick, param_get_uint,
 MODULE_PARM_DESC(spl_taskq_kick,
 	"Write nonzero to kick stuck taskqs to spawn more threads");
 
+/*
+ * This callback will be called exactly once for each core that comes online,
+ * for each dynamic taskq.  There are two kinds of dynamic taskqs that we
+ * attempt to expand; those who have TASKQ_THREADS_CPU_PCT set, and those
+ * created with the same number of threads as we have cores. In the CPU_PCT
+ * case, we need to redo the percentage calculation each time this is
+ * called. For the threads == cores case, we always want to create a new
+ * thread to keep the number of cores and threads in sync.
+ */
+static int
+spl_taskq_expand(unsigned int cpu, struct hlist_node *node)
+{
+	taskq_t *tq = list_entry(node, taskq_t, tq_hp_cb_node);
+	unsigned long flags;
+	int err = 0;
+
+	ASSERT(tq);
+	spin_lock_irqsave_nested(&tq->tq_lock, flags, tq->tq_lock_class);
+
+	if (!(tq->tq_flags & TASKQ_ACTIVE))
+		goto out;
+
+	if (tq->tq_flags & TASKQ_THREADS_CPU_PCT) {
+		int nthreads = MIN(tq->tq_orig_maxthreads, 100);
+		nthreads = MAX(((num_online_cpus() + 1) * nthreads) / 100, 1);
+		tq->tq_maxthreads = nthreads;
+	} else {
+		tq->tq_maxthreads++;
+	}
+
+	if (!((tq->tq_flags & TASKQ_DYNAMIC) && spl_taskq_thread_dynamic)) {
+		taskq_thread_t *tqt = taskq_thread_create(tq);
```

Comment:
It looks like we create a new thread even if tq_maxthreads didn't change (for non-DYNAMIC taskq's).  That doesn't seem right.